tools/codegen/gen_backend_stubs.py

```diff
@@ -40,34 +46,84 @@ def parse_backend_yaml(
     supported_autograd = yaml_values.pop('autograd', [])
     assert isinstance(supported, list), f'expected "autograd" to be a list, but got: {supported_autograd}'
 
+    if len(supported_autograd) > 0:
+        autograd_key = DispatchKey.try_parse(f'Autograd{backend}')
+        assert autograd_key is not None, f'The "autograd" key was specified, which indicates that you would like to override \
+the behavior of autograd for some operators on your backend. However "Autograd{backend}" is not a valid DispatchKey. \
+The set of valid dispatch keys is: {", ".join(k for k, v in DispatchKey.__members__.items())}'
+
     assert len(yaml_values.keys()) == 0, \
         f'{backend_yaml_path} contains unexpected keys: {", ".join(yaml_values.keys())}. \
 Only the following keys are supported: {", ".join(valid_keys)}'
 
     metadata: Dict[OperatorName, ExternalBackendMetadata] = {}
     for op in supported:
         op_name = OperatorName.parse(op)
-        m = ExternalBackendMetadata(op_name, backend, is_autograd=False)
+        m = ExternalBackendMetadata(op_name, is_autograd=False)
         metadata[m.operator] = m
     for op in supported_autograd:
         op_name = OperatorName.parse(op)
-        m = ExternalBackendMetadata(op_name, backend, is_autograd=True)
+        m = ExternalBackendMetadata(op_name, is_autograd=True)
         metadata[m.operator] = m
 
     native_functions_map: Dict[OperatorName, NativeFunction] = {
         f.func.name: f
         for f in concatMap(lambda f: [f] if isinstance(f, NativeFunction) else list(f.functions()), grouped_native_functions)
     }
 
+    def kernel_name(func: FunctionSchema) -> str:
+        # For external backends, we enforce that their names and signatures match the dispatcher convention
+        return dispatcher.name(func)
+
     def native_to_external(
             g: Union[NativeFunction, NativeFunctionsGroup]
     ) -> Union[ExternalBackendFunction, ExternalBackendFunctionsGroup]:
         if isinstance(g, NativeFunction):
             f = g
             m = metadata.get(f.func.name, None)
-            return ExternalBackendFunction(f, m)
+            dispatch_key = DispatchKey.parse(f'Autograd{backend}') \
+                if m is not None and m.is_autograd else DispatchKey.parse(backend)
+            kernel = kernel_name(f.func)
+            return ExternalBackendFunction(NativeFunction.with_dispatch_entry(f, dispatch_key, kernel), dispatch_key, m)
         elif isinstance(g, NativeFunctionsGroup):
-            return ExternalBackendFunctionsGroup.from_function_group(g, metadata)
+            out_meta = metadata.get(g.out.func.name, None)
+            kernel = kernel_name(g.out.func)
+            dispatch_key = DispatchKey.parse(f'Autograd{backend}') \
+                if out_meta is not None and out_meta.is_autograd else DispatchKey.parse(backend)
+            out = ExternalBackendFunction(NativeFunction.with_dispatch_entry(g.out, dispatch_key, kernel), dispatch_key, out_meta)
```

Comment:
The three blocks here feel juuuust long enough to want some deduping

Reply:
fair 😛